src/main/java/com/compilercharisma/chameleonbusinessstudio/service/AuthenticationService.java

```diff
@@ -40,30 +46,78 @@ public AbstractUser getLoggedInUser(){
             throw new UserNotRegisteredException();
         }
         OAuth2User googleUser = getLoggedInGoogleUser();
-        return users.get(googleUser.getAttribute("email"));
+        return googleUserToAbstractUser(googleUser);
     }
 
     public OAuth2User getLoggedInGoogleUser(){
         if(!isUserLoggedIn()){
             throw new NoUserLoggedInException();
         }
         Object p = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
-        if(!(p instanceof OAuth2User)){
-            throw new RuntimeException();
-        }
-        return (OAuth2User)p;
+        return principalToGoogleUser(p);
+    }
+
+    /**
+     * This method may replace some of our other methods, as it sounds like
+     * SecurityContextHolder is unreliable in Webflux: 
+     * https://stackoverflow.com/q/70050719
+     * 
+     * @return a mono containing the logged in user, or an error if no user is
+     *  logged in
+     */
+    public Mono<UserEntity> getLoggedInUserReactive(){
+        return ReactiveSecurityContextHolder
+            .getContext()
+            .map(SecurityContext::getAuthentication)
+            .handle((auth, sink)->{
+                // https://stackoverflow.com/a/53596358
+                if(isAuthenticatedUser(auth)){
+                    sink.next(auth.getPrincipal());
+                } else {
+                    sink.error(new NoUserLoggedInException());
+                }
+            })
+            .map(this::principalToGoogleUser)
+            .map(this::googleUserToAbstractUser)
+            .map(this::abstractUserToUserEntity);            
```

Comment:
The thing about this though, is that for users we are no longer using the UserEntity, or the AbstractEntity because they are complicated for Spring Security to figure by their type if they have necessary roles to do X activity.

Reply:
Wait... we're changing the user class now? I'm cool with us removing the AbstractEntity classes, but a lot of stuff depends on UserEntity, so do I have to change all of that as well?

Reply:
OK, it looks like I got most of the user changes, but I can't push yet